### ISSUE TITLE
`Layout/SpaceInsideArrayLiteralBrackets` is applied when `Layout/SpaceInsideReferenceBrackets` should be used

### ISSUE THREAD

**Issue by 4ndv** [rubocop/rubocop#14456]
## Expected behavior

When pattern matching block has a match arm like `in SomeClass[:some_thing]`, rubocop should use `Layout/SpaceInsideReferenceBrackets` instead of `Layout/SpaceInsideArrayLiteralBrackets`

## Actual behavior

Rubocop (incorrectly, imo) applies `Layout/SpaceInsideArrayLiteralBrackets` rule, producing results in conflict with `Layout/SpaceInsideReferenceBrackets`:

```ruby
def check_result(result)
  case result
  in Failure[ :some_error ] # <- Layout/SpaceInsideArrayBrackets is applied, putting the spaces between brackets
    Failure[:other_error] # <- Layout/SpaceInsideReferenceBrackets is applied, removing the spaces between brackets
  else
    result
  end
end
```

## Steps to reproduce the problem

I'm using `rubocop-rails-omakase` preset with minor changes, these two rules and code block from above should be enough to reproduce:

```yaml
# Use `[ a, [ b, c ] ]` not `[a, [b, c]]`
# Use `[]` not `[ ]`
Layout/SpaceInsideArrayLiteralBrackets:
  Enabled: true
  EnforcedStyle: space
  EnforcedStyleForEmptyBrackets: no_space

# Use `hash[:key]` not `hash[ :key ]`
Layout/SpaceInsideReferenceBrackets:
  Enabled: true
```

Related issue: #14144

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
$ [bundle exec] rubocop -V
1.79.2 (using Parser 3.3.9.0, Prism 1.4.0, rubocop-ast 1.46.0, analyzing as Ruby 3.4, running on ruby 3.4.2) [x86_64-linux]
  - rubocop-factory_bot 2.27.1
  - rubocop-rspec 3.6.0
  - rubocop-rspec_rails 2.31.0
  - rubocop-performance 1.25.0
  - rubocop-rails 2.33.3
```

**Comment by koic** [comment]
I think this is a configuration issue. Both `Layout/SpaceInsideArrayLiteralBrackets` and `Layout/SpaceInsideReferenceBrackets` cops should be made consistent with either `EnforcedStyle: space` or `EnforcedStyle: no_space`.

In `case 42; in Const[item]; end` the `Const[item]` is not index access but an array pattern matching. This can be confirmed by running `ruby --dump=parsetree`.

```console
$ ruby -v                                                                 
ruby 3.4.5 (2025-07-16 revision 20cda200d3) +PRISM [arm64-darwin24]

$ ruby --dump=parsetree -e 'case 42; in Const[item]; end'
@ ProgramNode (location: (1,0)-(1,28))
+-- locals: [:item]
+-- statements:
    @ StatementsNode (location: (1,0)-(1,28))
    +-- body: (length: 1)
        +-- @ CaseMatchNode (location: (1,0)-(1,28))
            +-- predicate:
            |   @ IntegerNode (location: (1,5)-(1,7))
            |   +-- IntegerBaseFlags: decimal
            |   +-- value: 42
            +-- conditions: (length: 1)
            |   +-- @ InNode (location: (1,9)-(1,23))
            |       +-- pattern:
            |       |   @ ArrayPatternNode (location: (1,12)-(1,23))
            |       |   +-- constant:
            |       |   |   @ ConstantReadNode (location: (1,12)-(1,17))
            |       |   |   +-- name: :Const
            |       |   +-- requireds: (length: 1)
            |       |   |   +-- @ LocalVariableTargetNode (location: (1,18)-(1,22))
            |       |   |       +-- name: :item
            |       |   |       +-- depth: 0
            |       |   +-- rest: nil
            |       |   +-- posts: (length: 0)
            |       |   +-- opening_loc: (1,17)-(1,18) = "["
            |       |   +-- closing_loc: (1,22)-(1,23) = "]"
            |       +-- statements: nil
            |       +-- in_loc: (1,9)-(1,11) = "in"
            |       +-- then_loc: nil
            +-- else_clause: nil
            +-- case_keyword_loc: (1,0)-(1,4) = "case"
            +-- end_keyword_loc: (1,25)-(1,28) = "end"
```

The same applies with parse.y instead of Prism.

```console
$ ruby --parser=parse.y --dump=parsetree -e 'case 42; in Const[item]; end'
###########################################################
## Do NOT use this node dump for any purpose other than  ##
## debug and research.  Compatibility is not guaranteed. ##
###########################################################

# @ NODE_SCOPE (id: 8, line: 1, location: (1,0)-(1,28))
# +- nd_tbl: :item
# +- nd_args:
# |   (null node)
# +- nd_body:
#     @ NODE_CASE3 (id: 7, line: 1, location: (1,0)-(1,28))*
#     +- nd_head:
#     |   @ NODE_INTEGER (id: 0, line: 1, location: (1,5)-(1,7))
#     |   +- val: 42
#     +- nd_body:
#     |   @ NODE_IN (id: 6, line: 1, location: (1,9)-(1,24))
#     |   +- nd_head:
#     |   |   @ NODE_ARYPTN (id: 4, line: 1, location: (1,12)-(1,22))
#     |   |   +- nd_pconst:
#     |   |   |   @ NODE_CONST (id: 1, line: 1, location: (1,12)-(1,17))
#     |   |   |   +- nd_vid: :Const
#     |   |   +- pre_args:
#     |   |   |   @ NODE_LIST (id: 3, line: 1, location: (1,18)-(1,22))
#     |   |   |   +- as.nd_alen: 1
#     |   |   |   +- nd_head:
#     |   |   |   |   @ NODE_LASGN (id: 2, line: 1, location: (1,18)-(1,22))
#     |   |   |   |   +- nd_vid: :item
#     |   |   |   |   +- nd_value:
#     |   |   |   |       (null node)
#     |   |   |   +- nd_next:
#     |   |   |       (null node)
#     |   |   +- rest_arg:
#     |   |   |   (null node)
#     |   |   +- post_args:
#     |   |       (null node)
#     |   +- nd_body:
#     |   |   @ NODE_BEGIN (id: 5, line: 1, location: (1,24)-(1,24))
#     |   |   +- nd_body:
#     |   |       (null node)
#     |   +- nd_next:
#     |       (null node)
#     +- case_keyword_loc: (1,0)-(1,4)
#     +- end_keyword_loc: (1,25)-(1,28)
```

For more information on array pattern matching, please refer to the following document:
https://speakerdeck.com/k_tsj/pattern-matching-new-feature-in-ruby-2-dot-7?slide=25

**Comment by 4ndv** [comment]
Thank you for detailed response, I completely forgot that we can do stuff like `in Failure[Symbol => pinned]`

Maybe pattern matching stuff can be spun off to a separate cop to allow configuring them differently from an array literals?
